workflow/controller/operator.go

```diff
@@ -3074,76 +3090,99 @@ func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
-		// If the controller is in reference mode, ensure that the stored spec is identical to the reference spec at every operation
-		wftSpec, err := woc.fetchWorkflowSpec()
+		err = woc.setExecWorkflow()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		if woc.wf.Status.StoredWorkflowSpec.String() != wftSpec.String() {
-			return nil, executionParameters, fmt.Errorf("workflowTemplateRef reference may not change during execution when the controller is in reference mode")
+	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
+		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+	} else {
+		err := woc.controller.setWorkflowDefaults(woc.wf)
+		if err != nil {
+			return err
 		}
+		woc.volumes = woc.wf.Spec.DeepCopy().Volumes
 	}
+	return nil
+}
+
+func (woc *wfOperationCtx) setStoredWFSpec() error {
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		var wftHolder wfv1.WorkflowSpecHolder
+		var err error
+		if woc.wf.Status.StoredWorkflowSpec == nil || woc.controller.Config.WorkflowRestrictions.MustNotChangeSpec() {
+			wftHolder, err = woc.fetchWorkflowSpec()
+			if err != nil {
+				return err
+			}
+		}
+		if woc.wf.Status.StoredWorkflowSpec == nil {
 
-	//In WorkflowTemplateRef scenario, we need to merge the  Workflow spec, :StoredWorkflowspec and Workflow Default for execWf.
-	// Overlay
-	targetWf := wfv1.Workflow{Spec: *woc.wf.Spec.DeepCopy()}
+			// Merge wfT Metadata to Wf
+			wfutil.MergeMetaDataTo(wftHolder.GetWorkflowMetadata(), &woc.wf.ObjectMeta)
 
-	err := wfutil.MergeTo(&wfv1.Workflow{Spec: *woc.wf.Status.StoredWorkflowSpec}, &targetWf)
-	if err != nil {
-		return nil, executionParameters, err
-	}
+			wfDefault := woc.controller.Config.WorkflowDefaults
+			if wfDefault == nil {
+				wfDefault = &wfv1.Workflow{}
+			}
+			// Merge wfDefault Metadata to Wf
+			wfutil.MergeMetaDataTo(&wfDefault.ObjectMeta, &woc.wf.ObjectMeta)
 
-	err = woc.controller.setWorkflowDefaults(&targetWf)
-	if err != nil {
-		log.WithFields(log.Fields{"key": woc.wf.Name, "error": err}).Warn("Failed to apply default workflow values")
-		return nil, executionParameters, err
-	}
-	// Setting the merged workflow to executable workflow
-	woc.execWf = &targetWf
+			// Merge WF spec WFT spec and WorkflowDefault spec
 
-	woc.volumes = woc.execWf.Spec.DeepCopy().Volumes
-	tmplRef := &wfv1.WorkflowStep{TemplateRef: woc.wf.Spec.WorkflowTemplateRef.ToTemplateRef(woc.execWf.Spec.Entrypoint)}
-	if len(woc.execWf.Spec.Arguments.Parameters) > 0 {
-		executionParameters.Parameters = util.MergeParameters(executionParameters.Parameters, woc.execWf.Spec.Arguments.Parameters)
-	}
-	executionParameters.Artifacts = util.MergeArtifacts(executionParameters.Artifacts, woc.execWf.Spec.Arguments.Artifacts)
+			mergedWf, err := wfutil.MergeWfSpecs(&woc.wf.Spec, wftHolder.GetWorkflowSpec(), &wfDefault.Spec)
```

Comment:
Merge WfSpec, WFTSpec, and WFDefaultSpec for `StoredWorkflowSpec`
order of precedence is WfSpec, WFTSpec and WFDefaultSpec